tests/components/nanoleaf/test_config_flow.py

```diff
@@ -0,0 +1,399 @@
+            {},
+        )
+    assert result4["type"] == "form"
+    assert result4["step_id"] == "link"
+    assert result4["errors"] == {"base": "not_allowing_new_tokens"}
+
+
+async def test_user_exception(hass: HomeAssistant) -> None:
+    """Test we handle Exception errors."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    with patch(
+        "homeassistant.components.nanoleaf.config_flow.Nanoleaf.authorize",
+        side_effect=Exception,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_HOST: TEST_HOST,
+            },
+        )
+    assert result2["type"] == "form"
+    assert result2["step_id"] == "user"
+    assert result2["errors"] == {"base": "unknown"}
+    assert not result2["last_step"]
+
+    with patch(
+        "homeassistant.components.nanoleaf.config_flow.Nanoleaf.authorize",
+        return_value=None,
+    ):
+        result3 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_HOST: TEST_HOST,
+            },
+        )
+    assert result3["step_id"] == "link"
+
+    with patch(
+        "homeassistant.components.nanoleaf.config_flow.Nanoleaf.authorize",
+        side_effect=Exception,
+    ):
+        result4 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {},
+        )
+    assert result4["type"] == "form"
+    assert result4["step_id"] == "link"
+    assert result4["errors"] == {"base": "unknown"}
+
+    with patch(
+        "homeassistant.components.nanoleaf.config_flow.Nanoleaf.authorize",
+        return_value=None,
+    ), patch(
+        "homeassistant.components.nanoleaf.config_flow.pynanoleaf_get_info",
+        side_effect=Exception,
+    ):
+        result5 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {},
+        )
+    assert result5["type"] == "abort"
+    assert result5["reason"] == "unknown"
+
+
+async def test_zeroconf_discovery(hass: HomeAssistant) -> None:
+    """Test zeroconfig discovery flow init."""
+    zeroconf = "_nanoleafms._tcp.local"
+    with patch(
+        "homeassistant.components.nanoleaf.config_flow.pynanoleaf_get_info",
+        return_value={"name": TEST_NAME},
+    ), patch(
+        "homeassistant.components.nanoleaf.config_flow.load_json",
+        return_value={},
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_ZEROCONF},
+            data={
+                "host": TEST_HOST,
+                "name": f"{TEST_NAME}.{zeroconf}",
+                "type": zeroconf,
+                "properties": {"id": TEST_DEVICE_ID},
+            },
+        )
+    assert result["type"] == "form"
+    assert result["step_id"] == "link"
```

Comment:
Please take each config flow test to either a create entry, error or abort result. We need to know that the whole flow works from end to end for each case.